File: DFS.cpp
#include <iostream>
#include <vector>
#include <climits>

class ChessBoard {
private:
    int size;
    std::vector<std::vector<bool>> visited;

public:
    ChessBoard(int n) : size(n), visited(n, std::vector<bool>(n, false)) {}

    bool isValid(int x, int y) {
        return x >= 0 && x < size && y >= 0 && y < size && !visited[x][y];
    }

    bool isOnBoundary(int x, int y) {
        return x == 0 || x == size - 1 || y == 0 || y == size - 1;
    }

    void markVisited(int x, int y) {
        visited[x][y] = true;
    }
};

int dx[] = { 2, 2, -2, -2, 1, 1, -1, -1 };
int dy[] = { 1, -1, 1, -1, 2, -2, 2, -2 };

/这里应用递归实现/
int DFS(ChessBoard& board, int x, int y, int steps) {
    if (board.isOnBoundary(x, y)) {
        return steps;
    }
    board.markVisited(x, y);
    int minSteps = INT_MAX;
    for (int i = 0; i < 8; i++) {
        int newX = x + dx[i];
        int newY = y + dy[i];
        if (board.isValid(newX, newY)) {
            int result = DFS(board, newX, newY, steps + 1);
            if (result != -1) {
                minSteps = std::min(minSteps, result);
            }
        }
    }
    return minSteps == INT_MAX ? -1 : minSteps;
}

int main() {
    int n;
    std::cout << "Enter the size of the chessboard (n x n): ";
    std::cin >> n;

    int startX, startY;
    std::cout << "Enter the starting position (x y): ";
    std::cin >> startX >> startY;

    ChessBoard board(n);
    int result = DFS(board, startX, startY, 0);
    if (result != -1) {
        std::cout << "The knight can reach the boundary in " << result << " steps from the given position." << std::endl;
    }
    else {
        std::cout << "The knight cannot reach the boundary from the given position." << std::endl;
    }

    return 0;
}

File: DFS.py
def is_valid(x, y, size, visited):
    return 0 <= x < size and 0 <= y < size and not visited[x][y]

def is_on_boundary(x, y, size):
    return x == 0 or x == size - 1 or y == 0 or y == size - 1

def DFS(x, y, size, visited, steps):
    if is_on_boundary(x, y, size):
        return steps

    min_steps = float('inf')
    visited[x][y] = True

    for i in range(8):
        newX, newY = x + dx[i], y + dy[i]
        if is_valid(newX, newY, size, visited):
            result = DFS(newX, newY, size, visited, steps + 1)
            if result != -1:
                min_steps = min(min_steps, result)

    visited[x][y] = False  # backtrack
    return min_steps if min_steps != float('inf') else -1

dx = [2, 2, -2, -2, 1, 1, -1, -1]
dy = [1, -1, 1, -1, 2, -2, 2, -2]

size = int(input("Enter the size of the chessboard (n x n): "))
startX, startY = map(int, input("Enter the starting position (x y): ").split())
visited = [[False] * size for _ in range(size)]
print(f"Using DFS, the knight can reach the boundary in {DFS(startX, startY, size, visited, 0)} steps.")
